File: 162/162_long.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int permutate(int num, int pos_1, int pos_2) {
    string num_str = to_string(num);
    swap(num_str[pos_1], num_str[pos_2]);
    return stoi(num_str);
}

int main() {
    string number;
    cin >> number;
    int k;
    cin >> k;
    vector<int> divisors = {5, 6, 10};
    vector<pair<int, int>> indices;
    for (int i = 0; i < number.size(); i++) {
        for (int j = i + 1; j < number.size(); j++) {
            indices.push_back({i, j});
        }
    }
    vector<int> array_nums = {stoi(number)};
    for (int i = 0; i < k; i++) {
        vector<int> new_array;
        for (int n = 0; n < array_nums.size(); n++) {
            for (auto pair : indices) {
                int new_num = permutate(array_nums[n], pair.first, pair.second);
                new_array.push_back(new_num);
            }
        }
        array_nums.clear();
        array_nums = new_array;
    }
    int counts = 0;
    for (auto item : array_nums) {
        for (auto divisor : divisors) {
            if (item % divisor == 0) {
                counts++;
                break;
            }
        }
    }
    long double probability = (long double)counts / (long double)array_nums.size();
    cout.precision(18);
    cout << probability << endl;
    return 0;
}

File: 162/162_long.py
import itertools

def permutate(num, pos_1, pos_2):
    num_array = list(str(num))
    num_array[pos_1], num_array[pos_2] = num_array[pos_2], num_array[pos_1]
    return int(''.join(num_array))

number = input()
k = int(input())

divisors = [5, 6, 10]
indices = list(itertools.combinations(range(len(number)), 2))

array_nums = [int(number)]
for i in range(k):
	new_array = []
	for n in range(len(array_nums)):
		for pair in indices:
			new_num = permutate(array_nums[n], pair[0], pair[1])
			new_array.append(new_num)
	array_nums.clear()
	array_nums.extend(new_array)

counts = 0
for item in array_nums:
	for divisor in divisors:
		if item % divisor == 0:
			counts += 1
			break

probability = counts / len(array_nums)
print(probability)
